nav2_collision_monitor/src/polygon.cpp

```diff
@@ -0,0 +1,376 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_collision_monitor/polygon.hpp"
+
+#include <exception>
+#include <utility>
+
+#include "geometry_msgs/msg/point.hpp"
+#include "geometry_msgs/msg/point32.hpp"
+
+#include "nav2_util/node_utils.hpp"
+
+#include "nav2_collision_monitor/kinematics.hpp"
+
+namespace nav2_collision_monitor
+{
+
+Polygon::Polygon(
+  const nav2_util::LifecycleNode::WeakPtr & node,
+  const std::string & polygon_name,
+  const std::shared_ptr<tf2_ros::Buffer> tf_buffer,
+  const std::string & base_frame_id,
+  const tf2::Duration & transform_tolerance)
+: node_(node), polygon_name_(polygon_name), action_type_(DO_NOTHING),
+  slowdown_ratio_(0.0), footprint_sub_(nullptr), tf_buffer_(tf_buffer),
+  base_frame_id_(base_frame_id), transform_tolerance_(transform_tolerance)
+{
+  RCLCPP_INFO(logger_, "[%s]: Creating Polygon", polygon_name_.c_str());
+}
+
+Polygon::~Polygon()
+{
+  RCLCPP_INFO(logger_, "[%s]: Destroying Polygon", polygon_name_.c_str());
+  poly_.clear();
+}
+
+
+bool Polygon::configure()
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  std::string polygon_topic, footprint_topic;
+
+  if (!getParameters(polygon_topic, footprint_topic)) {
+    return false;
+  }
+
+  if (!footprint_topic.empty()) {
+    footprint_sub_ = std::make_unique<nav2_costmap_2d::FootprintSubscriber>(
+      node, footprint_topic, *tf_buffer_,
+      base_frame_id_, tf2::durationToSec(transform_tolerance_));
+  }
+
+  if (visualize_) {
+    // Fill polygon_ points for future usage
+    std::vector<Point> poly;
+    getPolygon(poly);
+    for (const Point & p : poly) {
+      geometry_msgs::msg::Point32 p_s;
+      p_s.x = p.x;
+      p_s.y = p.y;
+      // p_s.z will remain 0.0
+      polygon_.points.push_back(p_s);
+    }
+
+    rclcpp::QoS polygon_qos = rclcpp::SystemDefaultsQoS();  // set to default
+    polygon_pub_ = node->create_publisher<geometry_msgs::msg::PolygonStamped>(
+      polygon_topic, polygon_qos);
+  }
+
+  return true;
+}
+
+void Polygon::activate()
+{
+  if (visualize_) {
+    polygon_pub_->on_activate();
+  }
+}
+
+void Polygon::deactivate()
+{
+  if (visualize_) {
+    polygon_pub_->on_deactivate();
+  }
+}
+
+std::string Polygon::getName() const
+{
+  return polygon_name_;
+}
+
+ActionType Polygon::getActionType() const
+{
+  return action_type_;
+}
+
+int Polygon::getMaxPoints() const
+{
+  return max_points_;
+}
+
+double Polygon::getSlowdownRatio() const
+{
+  return slowdown_ratio_;
+}
+
+double Polygon::getTimeBeforeCollision() const
+{
+  return time_before_collision_;
+}
+
+void Polygon::getPolygon(std::vector<Point> & poly) const
```

Comment:
This seems strange, why not return `std::vector<Point>` of polygon? Or a reference to it?

Reply:
This is made just for performance reasons. If we will return std::vector, it will be allocated somewhere outside the function to copy to. In the case if it is reference, it will be just a pointer to a some memory already allocated for `Polygon` or `Circle` object. We have no unnecessary copy of vector. That is why, I've chosen returning of reference to it.